File: app/src/main/java/com/example/chadrick/datalabeling/Models/DownloadTaskManager.java
package com.example.chadrick.datalabeling.Models;

import android.content.Context;
import com.example.chadrick.datalabeling.DownloadTask;
import java.util.HashMap;
import java.util.function.Consumer;

/**
 * Created by chadrick on 17. 10. 28.
 */

public class DownloadTaskManager {

  // this class should be a singleton.
  private static volatile DownloadTaskManager singleinstance = new DownloadTaskManager();



  // a hashmap to keep a reference for each ds id.

  private HashMap<DataSet, DownloadTask> downloadTaskHashMap = new HashMap<>();

  private DownloadTaskManager(){}

  public static DownloadTaskManager getInstance(){
    return singleinstance;
  }

  public synchronized boolean isAlreadyRegistered(DataSet dataSet){
    // for the given dataset, checks if it already has a not yet finished downloadtask registered to it.
    // it such a downloadtask exists, then return true
    // otherwise, return false

    if(downloadTaskHashMap.containsKey(dataSet)){
      DownloadTask fetchedDownloadTask = downloadTaskHashMap.get(dataSet);
      if(!fetchedDownloadTask.isFinished()){
        return true;
      }
    }

    return false;
  }

  public synchronized void addDownloadTask(DataSet dataSet, DownloadTask downloadTask){
    // check if a downloadtask for this dataset exists already.
    // if it exists, then check if it is finished. if finished, then delete it
    if(downloadTaskHashMap.containsKey(dataSet)){
      DownloadTask fetchedDownloadTask = downloadTaskHashMap.get(dataSet);
      if(downloadTask!=null){
        if(downloadTask.isFinished()){
          downloadTaskHashMap.remove(dataSet);
        }
      }
    }

    // add the new downloadtask, dataset pair to the hashmap
    downloadTaskHashMap.put(dataSet,downloadTask);

  }

}

File: app/src/main/java/com/example/chadrick/datalabeling/Models/DownloadTaskManager.kt
package com.example.chadrick.datalabeling.Models

import com.example.chadrick.datalabeling.Tasks.dszipDownloadTask

/**
 * Created by chadrick on 17. 12. 4.
 */

class DownloadTaskManager private constructor(){


    private val hashmap : HashMap<DataSet, dszipDownloadTask> = HashMap()
    private object holder { val INSTANCE = DownloadTaskManager()}

    companion object {
        val instance : DownloadTaskManager = holder.INSTANCE
    }

    @Synchronized fun isAlreadyRegistered(ds : DataSet) : Boolean {
        if(hashmap.containsKey(ds)){
            val fetchedtask = hashmap.get(ds)
            if( !(fetchedtask!!.isFinished())) return true
        }
        return false
    }

    @Synchronized fun addDownloadTask(ds: DataSet , param_downloadtask : dszipDownloadTask){
        if(hashmap.containsKey(ds)){
            // if exists and is finished, remove it.
            val task = hashmap.get(ds)
            if(task?.isFinished() ?: false){
                hashmap.remove(ds)
            }
        }

        hashmap.put(ds,param_downloadtask)
    }

    @Synchronized fun remove(ds:DataSet){
        if(hashmap.containsKey(ds)){
            val task = hashmap.get(ds)
            if(task?.isFinished() ?: true == false){
                hashmap.remove(ds)
                task?.cancel(true)
            }
        }
    }

}